# Calculus and Python

## Motivation
I am starting the last course in a unified treatment of Calculus at a local community college. My learning process includes reading each chapter twice, creating notes, and solving end-of-section problems. End-of-Section problems increase in difficulty where having the solution manual is neccessary. But some problems don't have the solution or require more exploration. The goal of this repository is to promote an experimental mindset when applying Calculus and Python with the help of chatbots. 

## How to use
The comments in the code will contain the relevant Calculus concepts for easy reference and clarity. But please use other resources to learn the underlying Calculus concepts fully. 

## Notation and Style
Solutions are given in a way that makes sense to me and will closely follow the textbook and solution manual. However, theorems, collolaries, formulas, etc. will use TI-BASIC and python in the comments.

## Citations
Swokowski, Earl W. *Calculus: The Classic Edition*. 7th ed., Brooks/Cole, 2012.

Swokowski, Earl W. *Student Solutions Manual for Calculus: The Classic Edition, Volume 2*. 7th ed., Brooks/Cole, 2012.

OpenAI. "ChatGPT". Accessed March 19, 2025. [https://www.openai.com/chatgpt].

# Install and Import

In [49]:
# Install the neccessary packages.
%pip install numpy sympy scipy pandas

# Import the neccessary libraries.
import numpy as np                      # For numerical operations and array handling.
import sympy as sp                      # For symbolic mathematics.
from scipy import integrate             # For numerical integration.
import pandas as pd                     # For data manipulation.

# Find Work Done
## Along y-axis from P1, P2
A constant force of magnitude 4 pounds has the same direction as the vector a = <i + j + k>. If distance is find the work done if the point of application moves along the y-axis from P1(0,2,0) to P2(0,-1,0).

In [50]:
# The magnitude of the constant force in pounds.
force_magnitude = 4

# Define and Solve displacement vector from inital, final points. Note inital, final applications of force.
inital_point = np.array([0, 2, 0])
final_point = np.array([0, -1, 0])
displacement_vector = final_point - inital_point

# Define the direction vector. For this problem the direction vector a equals <1i + 1j + 1k>.
a = np.array([1, 1, 1]) 

# The Unit vector in the direction of as defined by u = [1/norm(a)](a).
u = [1/np.linalg.norm(a)] * a

# The Force vector equals 4a/norm(a).
force_vector = force_magnitude * u 

# Work done, W is the dot product of the Force vector and the Displacement Vector.
W = np.dot(force_vector, displacement_vector)

# Output the results
print("Force Vector (F):", force_vector)
print("Displacement Vector (d):", displacement_vector)
print("Work Done in pounds (W):", W)

Force Vector (F): [2.30940108 2.30940108 2.30940108]
Displacement Vector (d): [ 0 -3  0]
Work Done in pounds (W): -6.9282032302755105


## Along a line from O to P
A constant force of magnitude 5 Newtons has the same direction as the positive z-axis. If distance is measured in meters, find the work done if the point of application moves along a line from the origin to the point P(1,2,3).

In [51]:
# The magnitude of the constant force in Newtons.
force_magnitude = 5

# Define and Solve displacement vector from inital, final points. Note inital, final applications of force.
initial_point = np.array([0, 0, 0])
final_point = np.array([1, 2, 3])
displacement_vector = final_point - initial_point

# Define the direction vector. For this problem the direction vector a equals <0i + 0j + 1k>.
a = np.array([0, 0, 1]) 

# The Unit vector in the direction of as defined by u = [1/norm(a)](a).
u = [1/np.linalg.norm(a)] * a 

# The Force vector equals 4a/norm(a).
force_vector = force_magnitude * u 

# Work done, W is the dot product of the Force vector and the Displacement Vector.
W = np.dot(force_vector, displacement_vector)

# Output the results
print("Force Vector (F):", force_vector)
print("Displacement Vector (d):", displacement_vector)
print("Work Done in Newtons (W):", W)

Force Vector (F): [0. 0. 5.]
Displacement Vector (d): [1 2 3]
Work Done in Newtons (W): 15.0


## Angle with Horizontal
A person pulls a wagon along level ground by exerting a force of 20 pounds on a handle that makes an angle of 30 degrees with the horizontal. Find the work done in pulling the wagon 100 feet.

In [52]:
# The magnitude of the constant force in pounds.
force_magnitude = 20

# Define and Solve displacement vector from inital, final points. Note inital, final applications of force.
initial_point = np.array([0, 0, 0])
final_point = np.array([100, 0, 0])
displacement_vector = final_point - initial_point

# Define the direction vector. The applied force makes a 30-degree angle with the horizontal.
# When force is applied at angle theta to the horizontal: x-component = cos(theta), y-component = sin(theta)
# When force is applied at angle theta to the vertical: y-component = cos(theta), x-component = sin(theta)
a = np.array([np.cos(np.radians(30)), np.sin(np.radians(30)), 0])

# The Unit vector in the direction of as defined by u = [1/norm(a)](a).
u = [1/np.linalg.norm(a)] * a 

# The Force vector equals 4a/norm(a).
force_vector = force_magnitude * u 

# Work done is the dot product of the force vector and displacement vector.
W = np.dot(force_vector, displacement_vector)

# Output the results
print("Force Vector (F):", force_vector)
print("Displacement Vector (d):", displacement_vector)
print("Work Done in Pounds (W):", W)

Force Vector (F): [17.32050808 10.          0.        ]
Displacement Vector (d): [100   0   0]
Work Done in Pounds (W): 1732.0508075688774


## Angle with x-axis and Incline
From the problem above, Force Vector is 17.32050808; Displacement Vector is <100,0,0>; Work done is 1732.0508075688774. Find the work done if the wagon is pulled, with the same force, 100 feet up an incline that makes an angle of 30 degrees with the horizontal.

In [53]:
# Define values from previous problem.
force_vector = np.array([17.32050808, 10, 0])
displacement_magnitude = 100

# Assign vector components by using the angle with the horizontal.
# When force is applied at angle theta to the horizontal: 
# x-component = cos(theta), y-component = sin(theta)
# When force is applied at angle theta to the vertical:
# y-component = cos(theta), x-component = sin(theta)
displacement_x = displacement_magnitude * np.cos(np.radians(30))
displacement_y = displacement_magnitude * np.sin(np.radians(30))
displacement_vector = np.array([displacement_x, displacement_y, 0])

# Work done is the dot product of the force vector and displacement vector.
W = np.dot(force_vector, displacement_vector)

# Output results:
print("Displacement Vector (d):", displacement_vector)
print("Work Done in Pounds (W):", W)

Displacement Vector (d): [86.60254038 50.          0.        ]
Work Done in Pounds (W): 2000.0000003733635


# Find the Angle between Vectors
## In Terms of Components
Given a sequence A-B-C-D of four bonded atoms, the angle between the plane formed by A, B, and C as well as the plane formed by B, C, and D is called the torsion angle theta of the bond. This torsion angle is used to explain the stability of molecular structures. If segment BC is placed along the z-axis, how can theta be computed in terms of the components of vector BA and vector CD?

In [54]:
# These points represent the position of four bonded atoms in an xyz-axis coordinate system.
# The problem states BC is aligned along the z-axis where i, j in B:<0, 0, 3> and C:<0, 0, 6> are zero.
A = np.array([1, 2, 3])
B = np.array([0, 0, 3])
C = np.array([0, 0, 6])
D = np.array([2, -1, 6])

# These vectors represent the bonds between the atoms.
BA = A - B
BC = C - B
CD = D - C

# The normal vector to a plane formed by two vectors (BA and BC) is given by:
# n1 = [a_y b_z - a_z b_y, a_z b_x - a_x b_z, a_x b_y - a_y b_x]
# Note: The cross product (or vector product) of vector a cross vector b of 
# vector a = <a1, a2, a3> and vector b = <b1, b2, b3> is vector a cross vector b = 
# matrix('a2 a3; b2 b3')i - matrix('a1 a3; b1 b3')j + matrix('a1 a2; b1 b2')k
# where i,j,k are unit vectors; can be used instead.
n1_x = BA[1] * BC[2] - BA[2] * BC[1]
n1_y = BA[2] * BC[0] - BA[0] * BC[2]
n1_z = BA[0] * BC[1] - BA[1] * BC[0]
n1 = np.array([n1_x, n1_y, n1_z])

# The above procedure applies here too.
n2_x = BC[1] * CD[2] - BC[2] * CD[1]
n2_y = BC[2] * CD[0] - BC[0] * CD[2]
n2_z = BC[0] * CD[1] - BC[1] * CD[0]
n2 = np.array([n2_x, n2_y, n2_z])

# The torsion angle can be calculated by the dot product over normals n1, n2.
# If theta is the angle between nonzero vectors n1 and n2, then 
# cos(theta) = DotP(n1, n2) / [Norm(n1)][Norm(n2)].
cos_theta = np.dot(n1, n2) / (np.linalg.norm(n1) * np.linalg.norm(n2))

# Solving for theta, then applying clip to avoid numerical errors.
theta = np.arccos(np.clip(cos_theta, -1.0, 1.0))
theta_degrees = np.degrees(theta)

# Display results.\
print("Coordinates:")
print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)
print()

print("Vectors:")
print("BA =", BA)
print("BC =", BC)
print("CD =", CD)
print()

print("Normal vector to plane ABC (n1):")
print("n1_x =", n1_x, "n1_y =", n1_y, "n1_z =", n1_z)
print("n1 =", n1)
print()

print("Normal vector to plane BCD (n2):")
print("n2_x =", n2_x, "n2_y =", n2_y, "n2_z =", n2_z)
print("n2 =", n2)
print()

print("Torsion angle in radians:")
print(theta)
print("Torsion angle in degrees:")
print(theta_degrees)

Coordinates:
A = [1 2 3]
B = [0 0 3]
C = [0 0 6]
D = [ 2 -1  6]

Vectors:
BA = [1 2 0]
BC = [0 0 3]
CD = [ 2 -1  0]

Normal vector to plane ABC (n1):
n1_x = 6 n1_y = -3 n1_z = 0
n1 = [ 6 -3  0]

Normal vector to plane BCD (n2):
n2_x = 3 n2_y = 6 n2_z = 0
n2 = [3 6 0]

Torsion angle in radians:
1.5707963267948966
Torsion angle in degrees:
90.0


# Volume of Box 
## Find the Angle between Vectors
Suppose vector a, vector b, and vector c represent sides of an oblique box. Show that abs(Dotp(CrossP(a,b),c)) is the volume of the box.

In [55]:
# Define example vectors.
a = np.array([3, 0, 0])
b = np.array([0, 4, 0])
c = np.array([0, 0, 5])

# Note: The cross product (or vector product) of vector a cross vector b of 
# vector a = <a1, a2, a3> and vector b = <b1, b2, b3> is vector a cross vector b = 
# matrix('a2 a3; b2 b3')i - matrix('a1 a3; b1 b3')j + matrix('a1 a2; b1 b2')k
# where i,j,k are unit vectors.
cross_product = np.cross(a, b)

# The scalar triple prodcut can be calculated by taking vector a cross vector b 
# dot c as shown in TI-Basic: DotP(CrossP(a,b),c)).
scalar_triple = np.dot(cross_product, c)

# The volume is the absolute value of the scalar triple product must be taken 
# as the angle may be an obtuse angle.
volume = abs(scalar_triple)

# Print the results.
print("Vector a:", a)
print("Vector b:", b)
print("Vector c:", c)
print()
print("Cross product:", cross_product)
print("Scalar triple product: ", scalar_triple)
print("Volume of the parallelepiped:", volume)

Vector a: [3 0 0]
Vector b: [0 4 0]
Vector c: [0 0 5]

Cross product: [ 0  0 12]
Scalar triple product:  60
Volume of the parallelepiped: 60


## Find the Volume of a Box
Using the above problem, find the volume of the box having adjacent sides AB, AC, and AD. A(2,1,-1), B(3,0,2), C(4,-2,1), D(5,-3,0).

In [56]:
# Given the points representing vertices of the box:
A = np.array([2, 1, -1])
B = np.array([3, 0, 2])
C = np.array([4, -2, 1])
D = np.array([5, -3, 0])

# THe volume of the box with adjacent vector sides AB, AC, and AD
AB = B - A
AC = C - A
AD = D - A

# The volume of the parallelepiped is defined by the absolute value of their 
# scalar triple product: abs(DotP(u,CrossP(v,w)) where u = AB, v = AC, and w = AD.
cross_product = np.cross(AC, AD)
scalar_triple = np.dot(AB, cross_product)
volume = abs(scalar_triple)

# Print results.
print("Point A:", A)
print("Point B:", B)
print("Point C:", C)
print("Point D:", D)
print()

print("Vector AB (B - A):", AB)
print("Vector AC (C - A):", AC)
print("Vector AD (D - A):", AD)
print()

print("Cross product of AC and AD:", cross_product)
print("Scalar triple product:", scalar_triple)
print("Volume of the parallelepiped (absolute value):", volume)

Point A: [ 2  1 -1]
Point B: [3 0 2]
Point C: [ 4 -2  1]
Point D: [ 5 -3  0]

Vector AB (B - A): [ 1 -1  3]
Vector AC (C - A): [ 2 -3  2]
Vector AD (D - A): [ 3 -4  1]

Cross product of AC and AD: [5 4 1]
Scalar triple product: 4
Volume of the parallelepiped (absolute value): 4
